docs/source/best_practices.rst

```diff
@@ -0,0 +1,187 @@
+you can often rewrite your code using visitor methods instead. Nested instance checks
+can instead be unrolled into simpler code using visitors with matcher decorators. For
+example, instead of writing the following::
+
+    class CountBazFoobarArgs(cst.CSTVisitor):
+        """
+        Given a set of function names, count how many arguments to those function
+        calls are the identifiers "baz" or "foobar".
+        """
+
+        def __init__(self, functions: Set[str]) -> None:
+            super().__init__()
+            self.functions: Set[str] = functions
+            self.arg_count: int = 0
+
+        def visit_Call(self, node: cst.Call) -> None:
+            # See if the call itself is one of our functions we care about
+            if isinstance(node.func, cst.Name) and node.func.value in self.functions:
+                # Loop through each argument
+                for arg in node.args:
+                    # See if the argument is an identifier matching what we want to count
+                    if isinstance(arg.value, cst.Name) and arg.value.value in {"baz", "foobar"}:
+                        self.arg_count += 1
+                        
+Consider writing instead::
+
+    class CountBazFoobarArgs(m.MatcherDecoratableVisitor):
+        """
+        Given a set of function names, count how many arguments to those function
+        calls are the identifiers "baz" or "foobar".
+        """
+
+        def __init__(self, functions: Set[str]) -> None:
+            super().__init__()
+            self.functions: Set[str] = functions
+            self.arg_count: int = 0
+            self.call_stack: List[str] = []
+
+        def visit_Call(self, node: cst.Call) -> None:
+            # Store all calls in a stack
+            if m.matches(node.func, m.Name()):
+                self.call_stack.append(cst.ensure_type(node.func, m.Name).value)
+
+        def leave_Call(self, original_node: cst.Call) -> None:
+            # Pop the latest call off the stack
+            if m.matches(node.func, m.Name()):
+                self.call_stack.pop()
+
+        @m.visit(m.Arg(m.Name("baz") | m.Name("foobar")))
+        def _count_args(self, node: cst.Arg) -> None:
+            # See if the most shallow call is one we're interested in, so we can
+            # count the args we care about only in calls we care about.
+            if self.call_stack[-1] in self.functions:
+                self.arg_count += 1
+
+While there is more code than the previous example, it is arguably easier to understand
+and maintain each piece of the code. Note that LibCST is traversing the tree completely
+in both cases, so while the first appears to be faster, it is actually doing the same
+amount of work as the second.
+
+Prefer ``updated_node`` when modifying trees
+--------------------------------------------
+
+When you are using :ref:`libcst-visitors` to modify a LibCST tree, only return
```

Comment:
Can we make it more explicit using `Transformer`? Only Transformer can return modifications in `leave_<Node>`.

Reply:
Good idea.